api/handlers.go

```diff
@@ -890,8 +890,47 @@ func getEvents(c *context, w http.ResponseWriter, r *http.Request) {
 		f.Flush()
 	}
 
-	c.eventsHandler.Add(r.RemoteAddr, w)
-	c.eventsHandler.Wait(r.RemoteAddr, until)
+	eventsChan, cancelFunc := c.watchQueue.Watch()
+	defer cancelFunc()
+
+	// create timer for --until
+	var (
+		timer   *time.Timer
+		timerCh <-chan time.Time
+	)
+	if until > 0 {
+		dur := time.Unix(until, 0).Sub(time.Now())
+		timer = time.NewTimer(dur)
+		timerCh = timer.C
+	}
+	var closeNotify <-chan bool
+	if closeNotifier, ok := w.(http.CloseNotifier); ok {
+		closeNotify = closeNotifier.CloseNotify()
+	}
+
+	for {
+		select {
+		case eChan := <-eventsChan:
+			if e, ok := eChan.(*cluster.Event); ok {
+				data, err := normalizeEvent(e)
+				if err != nil {
+					return
+				}
+				_, err = fmt.Fprint(w, string(data))
```

Comment:
Why not write directly here, rather than reallocate the data as a string? `w.Write(data)` should be sufficient.